### PR TITLE
torch/monitor: TensorboardEventHandler

### PR DESCRIPTION
**Pull request by d4l3k**
Summary: This adds the beginnings of a TensorboardEventHandler which will log stats to Tensorboard.

Test Plan: buck test //caffe2/test:monitor

Differential Revision: D33719954

### CODE REVIEW

File: test/test_monitor.py [hunk 1]
```diff
@@ -98,6 +100,60 @@ def handler(event: Event) -> None:
...
         self.assertEqual(len(events), 2)
 
+TEST_TENSORBOARD = True
+try:
+    from torch.utils.tensorboard import SummaryWriter
+    from tensorboard.backend.event_processing import (
+        plugin_event_multiplexer as event_multiplexer,
+    )
+except ImportError:
+    TEST_TENSORBOARD = False
```
Comment (edward-io) [inline test/test_monitor.py:110]: can we put this in setUp?

File: test/test_monitor.py [hunk 2]
```diff
@@ -98,6 +100,60 @@ def handler(event: Event) -> None:
...
+
+class TestMonitorTensorboard(TestCase):
+    def setUp(self):
+        if not TEST_TENSORBOARD:
+            return self.skipTest("Skip the test since TensorBoard is not installed")
+        self.temp_dirs = []
+
+    def createSummaryWriter(self):
```
Comment (edward-io) [inline test/test_monitor.py:118]: nit: should we stick to snake case for our own methods?

File: test/test_monitor.py [hunk 3]
```diff
@@ -98,6 +100,60 @@ def handler(event: Event) -> None:
...
+        temp_dir = tempfile.TemporaryDirectory()  # noqa: P201
+        self.temp_dirs.append(temp_dir)
+        return SummaryWriter(temp_dir.name)
+
+    def tearDown(self):
+        super().tearDown()
```
Comment (edward-io) [inline test/test_monitor.py:124]: nit: the parent class doesn't do anything for teardown

File: test/test_monitor.py [hunk 4]
```diff
@@ -98,6 +100,60 @@ def handler(event: Event) -> None:
...
+        # Remove directories created by SummaryWriter
+        for temp_dir in self.temp_dirs:
+            temp_dir.cleanup()
+
+    def test_event_handler(self):
+        with self.createSummaryWriter() as w:
```
Comment (edward-io) [inline test/test_monitor.py:130]: why do we use a context manager here?

File: test/test_monitor.py [hunk 5]
```diff
@@ -98,6 +100,60 @@ def handler(event: Event) -> None:
...
+                2,
+            )
+            for i in range(10):
+                s.add(i)
+            self.assertEqual(s.count, 0)
+
+            unregister_event_handler(handle)
+
+        mul = event_multiplexer.EventMultiplexer()
+        mul.AddRunsFromDirectory(self.temp_dirs[-1].name, None)
```
Comment (edward-io) [inline test/test_monitor.py:145]: the second argument is None by default

File: test/test_monitor.py [hunk 6]
```diff
@@ -98,6 +100,60 @@ def handler(event: Event) -> None:
...
+        mul.Reload()
+        scalar_dict = mul.PluginRunToTagToContent("scalars")
+        raw_result = {
+            tag: mul.Tensors(run, tag)
+            for run, run_dict in scalar_dict.items()
+            for tag in run_dict
+        }
+        self.assertCountEqual(raw_result.keys(), {"asdf.sum", "asdf.count"})
+        for v in raw_result.values():
+            self.assertEqual(len(v), 5)
```
Comment (edward-io) [inline test/test_monitor.py:155]: should we test the actual value here too?

File: torch/monitor/__init__.py [hunk 7]
```diff
@@ -1 +1,32 @@
...
+if TYPE_CHECKING:
+    from torch.utils.tensorboard import SummaryWriter
+
+
+STAT_EVENT = "torch.monitor.Stat"
+
+class TensorboardEventHandler:
+    """
+    TensorboardEventHandler is an event handler that will write known events to
+    the provided SummaryWriter.
```
Comment (edward-io) [inline torch/monitor/__init__.py:14]: we should also mention it only handles scalars

**Review by edward-io — APPROVED** [review]
LGTM, some nits

File: test/test_monitor.py (hunk 1 above)
Comment (d4l3k) [reply test/test_monitor.py:110]: this was copied from the tensorboard tests, but done

File: torch/monitor/__init__.py (hunk 7 above)
Comment (edward-io) [reply torch/monitor/__init__.py:14]: did we update this? :)

**Review by edward-io — APPROVED** [review]
lgtm, tiny nit

File: torch/monitor/__init__.py (hunk 7 above)
Comment (d4l3k) [reply torch/monitor/__init__.py:14]: see the next line :)

**Comment by d4l3k** [comment]
mypy failing due to trunk breakage https://www.torch-ci.com/

https://github.com/pytorch/pytorch/commit/bbe6144b453c7d990bf8c9fb184d14bad30a457d

**Pull request closed without merging**